### PR TITLE
Ensure width of table is not wider than console #6

### PR DESCRIPTION
**Pull request by zumo-sk**
- Addressing #6 
- Adding a possibility to align table on the output canvas

### CODE REVIEW

**Comment by zumo-sk** [comment]
Hi Tom,

If you'd have time, I'd appreciate review and comments. There are few things I was considering adding and reworking (for example, using a separate internal property for marking columns as hidden instead of using Hide property), but all in all, I am satisfied with the outcome.
Also, I don't really see the non-covered method list, is build system providing it? Or should I just use sixth sense? 😅 
Any comments or suggestions are welcome. And happy new year!

Marek

**Comment by tdwright** [comment]
Hi @zumo-sk - happy new year! 🍾 

Thanks for doing all this work. From a superficial glance it looks really good. I'll probably get some time to review it properly over the weekend.

With regards to coverage, I think I'd forgotten that it doesn't work for PRs. (Think it's a security thing?) I'll run it locally when I review and will let you know.

Tom 😄

File: ConTabs/Table.cs [hunk 1]
```diff
@@ -110,6 +122,9 @@ private Table()
             TableStyle = Style.Default;
             HeaderAlignment = Alignment.Default;
             ColumnAlignment = Alignment.Default;
+            TableAlignment = Alignment.Default;
+            CanvasWidth = Console.WindowWidth - 1; // it would word wrap
```
Comment (tdwright) [inline ConTabs/Table.cs]: Binding to the System.Console class worries me. What happens when someone wants to use ConTabs outside of the console?

So far we've been able to ignore the Console by treating everything as a simple string and leaving it up to our downstream consumers to work out what to do with it. On the other hand, the system console is probably where ConTabs gets called 99% of the time, so there is merit in making some assumptions. It's a tricky one!

Two options spring to mind:

1. Use a new class (called e.g. `TableContainer`) that wraps a `Table` object and is coupled to Console with this sort of code. The the container could create the table, set the width based on the console width, and render it to the console directly.

2. Add a helper method somewhere that sets the `CanvasWidth` to `Console.WindowWidth - 1`.

3. Add some logic that tries to detect whether we're in a Console (maybe [like this?](https://stackoverflow.com/a/15268625/50151)) and set the `CanvasWidth` to the console width or a sensible default.

Each have their own advantages and disadvantages. But I think it is important to retain our "portability" - even if its only for the 1% of users who run ConTabs in more exotic contexts.

PS After a bit more thought, I think those three options are in my order of preference. Still keen to hear your thoughts though.

File: ConTabs/Column.cs [hunk 2]
```diff
@@ -54,21 +54,6 @@ public class Column
         /// A List of the values stored
         /// </summary>
         public List<Object> Values { get; set; }
-        public int MaxWidth
```
Comment (tdwright) [inline ConTabs/Column.cs:56]: Does removing this mean that the table will always fill the full width provided to it? I think it would be better if we retained the default behaviour of taking as much room as needed, for two reasons:

1) If we don't need the full width of the console, we shouldn't stretch to use it.

2) Sometimes we will want to just render a table without having to know what width we want it to be.

The advantage of the current behaviour is that it's really simple to knock out a quick table in almost any environment. Being able to set a max width (or having some method work it out from the console window) is great, but it should _add to_ rather than _replace_ the default behaviour.

File: ConTabs/OutputBuilder.cs [hunk 3]
```diff
@@ -50,6 +60,11 @@ private OutputBuilder(Table<T2> t, Style s)
                     InsertVerticalPadding(table.Padding.Bottom, style.Wall); NewLine();
                 }
                 HLine(TopMidBot.Bot);
+
+                foreach (var col in hiddenCols)
+                {
+                    col.Hide = false;
```
Comment (tdwright) [inline ConTabs/OutputBuilder.cs]: I think this is the main reason I'd advocate using a separate property for this sort of "hiding". When we're done rendering the table, we want the columns we've hidden due to space constraints to be unhidden, but we don't want this for the ones where the _user_ did `.Hide = true`.

File: ConTabs/OutputBuilder.cs [hunk 4]
```diff
@@ -140,6 +156,44 @@ private void DataLine(List<CellParts> parts, int line)
...
+            private int GetOptimalTableWidth(List<Column> hiddenCols)
+            {
+                int tableWidth = GetTableWidthAfterApplyingStretchStyles();
+
+                while (table.CanvasWidth < tableWidth && table._colsShown.Count > 0)
+                {
+                    var columnToHide = table._colsShown[table._colsShown.Count - 1];
+                    hiddenCols.Add(columnToHide);
+                    tableWidth -= columnToHide.LongStringBehaviour.DisplayWidth + table.Padding.GetHorizontalPadding() + 1;
+                    columnToHide.Hide = true;
```
Comment (tdwright) [inline ConTabs/OutputBuilder.cs]: As you allude to in your PR, re-using the `Hide` property doesn't quite feel right. Hard to put my finger on it, but it feels like we're doing something different here. Perhaps we could add an internal property called "dropped" or "suppressed", or something?

File: ConTabs/OutputBuilder.cs [hunk 5]
```diff
@@ -140,6 +156,44 @@ private void DataLine(List<CellParts> parts, int line)
...
+                    int colWidth = table.TableStretchStyles.CalculateOptimalWidth(column);
+                    column.LongStringBehaviour.DisplayWidth = colWidth;
+                    colWidths += colWidth;
+                }
+                int realWidth = colWidths + GetWidthOfPaddingAndBorders();
+                int adaptedWidth = table.TableStretchStyles.CalculateAdditionalWidth(table._colsShown, realWidth, table.CanvasWidth);
+                return adaptedWidth + GetWidthOfPaddingAndBorders();
+            }
+
+            private int GetOptimalTableWidth(List<Column> hiddenCols)
```
Comment (tdwright) [inline ConTabs/OutputBuilder.cs]: Passing `hiddenCols` as an argument here feels a bit off. This method is an instance member of the `Table` class, so it can access the list of columns directly, without needing to pass around additional lists.

**Review by tdwright — CHANGES_REQUESTED** [review]
Hi @zumo-sk,

Sorry it took me longer than I'd said to review this. I kept running out of time!

Anyway, I think you've made a great start here. I've put my thoughts inline in the code as review comments, but my overall feeling is that we need to rethink a couple of aspects. The first (which you mention in the PR) is the use of the existing `Hide` property, which I think is causing more trouble than it's saving. 😟

The second theme of my feedback is around making this functionality additional to the existing behaviour, rather than supplanting it. I'm keen that we ensure that ConTabs continues to be highly portable and simple to implement. Adding new features without breaking an existing API is always a challenge, but that's what makes it interesting, right? 😉 

Please don't be disheartened by the feedback. I know you share my commitment to making ConTabs great and I really appreciate the effort you've put in to this PR. A few changes and it'll be 💯 ready to go.

Cheers,
Tom

**Comment by tdwright** [comment]
Oh, also, I've changed the merge target from `master` to `develop`. This introduced a small conflict, due to the fact that `_colsShown` is now called `ColsShown`. Sorry!

File: ConTabs/Table.cs (hunk 1 above)
Comment (zumo-sk) [reply ConTabs/Table.cs]: The idea was to set the default values from console width - and if users wants, they can override the CanvasWidth property - as done for example in some of the tests, and as we can do with other table properties. But if you insist it does not belong to Table class, but should be in a container, I am willing to try to implement it in that way. It's your project, I'm just trying to help! 😄

File: ConTabs/Column.cs (hunk 2 above)
Comment (zumo-sk) [reply ConTabs/Column.cs:56]: I seem to remember that the behavior stayed the same, only difference is that this calculation is being used only if you use some of the stretch styles. You can see that the default behavior stayed unchanged for example from tests - even when they are using the long strings, the outputted tables are still the same than before my changes. I was thinking about backwards compatibility, but I could not think of real case scenario where it could have become a problem. But if you disagree, I am willing to think it through.

File: ConTabs/OutputBuilder.cs (hunk 3 above)
Comment (zumo-sk) [reply ConTabs/OutputBuilder.cs]: It has been some time since I wrote it, but I think this is a naming issue. Variable hiddenCols does not represent all the columns that are hidden, just the columns that were hidden because of not being able to fit on the canvas. Because of that, it will not show the columns that were hidden manually by user. But just from this confusion, I agree that using Hide property was not a good idea and I should have used another property for that (probably internal, not a public one), that will be used just to suppress the columns if they can't fit.

File: ConTabs/OutputBuilder.cs (hunk 4 above)
Comment (zumo-sk) [reply ConTabs/OutputBuilder.cs]: Agreed. Will do.

File: ConTabs/OutputBuilder.cs (hunk 5 above)
Comment (zumo-sk) [reply ConTabs/OutputBuilder.cs]: I think this is the problem that I described earlier - hiddenCols is just a local variable that is being filled with columns that could not fit here. But maybe it would make sense to make it into a Table property without actually passing the reference.

**Comment by zumo-sk** [comment]
Thanks for the review! I think I added comments to all your comments. I am going to travel for few days again tomorrow, but after I return, I'll take a look again and fix what I am sure needs fixing 😄

**Comment by zumo-sk** [comment]
Hi Tom,
Sorry for a delay, I have been pretty busy. But I managed to travel more than I originally planned and I can only say that Finland and Estonia are lovely in the winter, but Ukraine is nicer in the summer. But back to business.
I tried to address everything from your review and even when I am not 100% sure whether not having default CanvasWidth set to console width is a good idea or not, I've removed it. In the program.cs everyone can see how to correctly set it up to work in console, so hopefully that's enough. I also addressed what was originally a corner case and now is a default value - having CanvasWidth = 0. All confirmed with new tests.
One more thing, I found another test that did not work in Danish locale settings (decimal comma) - so I fixed the test in the same matter as I fixed the ones before. 😉
Looking forward for new challenges raised by review comments! 😅
Marek

**Comment by zumo-sk** [comment]
Oh, one more thing. I am not sure about the workflow here. If I have to address review comments one by one with code snippets, let me know and I'll do it.

**Comment by tdwright** [comment]
Hi @zumo-sk,

Thanks for these changes. Sounds like you've given it plenty of thought. I'll review them properly ASAP. I'm at a conference this week, but might get a chance at the airport? We'll see! In the grand scheme of things, this is great timing, as I was planning on tackling #46 myself soon and can see how it might layer on top of your work.

Sounds like you've been busy travelling. Work? Or holidays? Do you enjoy snow sports at all? Skiing is one of the things I really miss about Europe.

Tom

**Comment by zumo-sk** [comment]
We'll see. In worst case you'll request some more changes and I'll try to implement them faster this time. 😄
Finland and Estonia were holidays, Ukraine was work (and I expect some more work related traveling soon). None of them included sports, but I would understand why would anyone go skiing there - there was plenty of snow and it was not crowded by tourists. I was never a ski person, but my wife was, and she's missing skiing and snowboarding since we moved to Denmark few years ago (you know, in a country with highest peak called Sky Mountain being 147m high, it's not very ski friendly). This time, the travels were mostly to enjoy local food (reindeer stew, anyone?), visit local board game cafes and to learn something about the countries we didn't really know that much of. And I have to say, mission was successful.
When did you actually moved to Australia? And was it because you missed poisonous animals, crazy temperatures,, or did you have other motivation? 😃

**Review by tdwright — APPROVED** [review]

**Comment by tdwright** [comment]
Well @zumo-sk - these changes are really nice. I'm going to merge them now. I still plan on rolling these changes into [v2](https://github.com/tdwright/contabs/milestone/1), which may be some time away, but I'll put out another preview release in the meantime. I spend a long time this evening playing with all the different options and I have to say I really like this functionality. Great job!

We moved to Australia last (northern-hemisphere) summer. I blogged about it [here](http://blog.tdwright.co.uk/2018/05/14/toms-going-down-under/), but to answer your question: no, it was certainly not for the poisonous animals. In fact, the main motivation was for my wife's work. Your holiday sounds lovely - I hope you had a relaxing time!

**Comment by tdwright** [comment]
Preview is up on NuGet: https://www.nuget.org/packages/ConTabs.tdwright/2.0.0-preview2

Thanks again @zumo-sk!
